### DISCUSSION TITLE
Thread deadlock

### DISCUSSION THREAD

**Discussion by luis-guideti** [Auties00/WhatsappWeb4j#356] [General]
Im opening this on Discussions because I dont know if its an issue that can be fixed given the blocking nature of the Socket itself, but maybe it should be documented?

I recently upgraded from 3.2.3 (which started banning the number when reading the qrcode) to 3.5.1 and on my computer with a single number everything worked fine, but as soon as I uploaded to a VPS that I have with 6 numbers it didnt work. I connect to the numbers in sequence with a delay between them, the first number connected fine, but as soon as the second connected they all get stuck, no stacktraces or anything, I would just get a socket OPEN event and some time later a CLOSE event with no other information.

I noticed all threads being stuck in the same place, so I increase the available cores from 4 to 8 (using -XX:ActiveProcessorCount=8) and it improved, but after some time they all got stuck again.

Now I setup a cached thread pool, overriding the shutdownNow() because it is called when the socket closes, I saw peaks of 10 active threads for 6 numbers mostly in idle. It appears it requires quite a few available threads to work, specialy since the CompletableFuture still uses the common pool even when provided with an executor.

`    
private static final ThreadPoolExecutor WHATSAPP_EXECUTOR;

    static {

        ThreadCounter factory = new ThreadCounter("WhatsApp Thread", true, Thread.NORM_PRIORITY);

        WHATSAPP_EXECUTOR = new ThreadPoolExecutor(
                0, 64,
                5, TimeUnit.MINUTES,
                new SynchronousQueue<>(),
                factory) {
                    @Override
                    public List<Runnable> shutdownNow() {
                        return null;
                    }
              };
    }`

Thread dump

[threads-1690653410405.txt](https://github.com/Auties00/Cobalt/files/12207713/threads-1690653410405.txt)

**Comment by Auties00** [comment]
> Im opening this on Discussions because I dont know if its an issue that can be fixed given the blocking nature of the Socket itself, but maybe it should be documented?
> 
> I recently upgraded from 3.2.3 (which started banning the number when reading the qrcode) to 3.5.1 and on my computer with a single number everything worked fine, but as soon as I uploaded to a VPS that I have with 6 numbers it didnt work. I connect to the numbers in sequence with a delay between them, the first number connected fine, but as soon as the second connected they all get stuck, no stacktraces or anything, I would just get a socket OPEN event and some time later a CLOSE event with no other information.
> 
> I noticed all threads being stuck in the same place, so I increase the available cores from 4 to 8 (using -XX:ActiveProcessorCount=8) and it improved, but after some time they all got stuck again.
> 
> Now I setup a cached thread pool, overriding the shutdownNow() because it is called when the socket closes, I saw peaks of 10 active threads for 6 numbers mostly in idle. It appears it requires quite a few available threads to work, specialy since the CompletableFuture still uses the common pool even when provided with an executor.
> 
> ` private static final ThreadPoolExecutor WHATSAPP_EXECUTOR;
> 
> ```
> static {
> 
>     ThreadCounter factory = new ThreadCounter("WhatsApp Thread", true, Thread.NORM_PRIORITY);
> 
>     WHATSAPP_EXECUTOR = new ThreadPoolExecutor(
>             0, 64,
>             5, TimeUnit.MINUTES,
>             new SynchronousQueue<>(),
>             factory) {
>                 @Override
>                 public List<Runnable> shutdownNow() {
>                     return null;
>                 }
>           };
> }`
> ```
> 
> Thread dump
> 
> [threads-1690653410405.txt](https://github.com/Auties00/Cobalt/files/12207713/threads-1690653410405.txt)


Can't you just use a different thread here?
https://github.com/Auties00/Cobalt/blob/d128d9aa893c244dd22027de0ed2678d845580c4/src/main/java/it/auties/whatsapp/api/OptionsBuilder.java#L114
This should make it work
